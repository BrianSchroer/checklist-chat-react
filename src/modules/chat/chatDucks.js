// Chat actions/reducer "ducks" file (https://github.com/erikras/ducks-modular-redux)

import initialState from '../../app/store/initialState';
import {beginAjaxCall, ajaxCallError} from '../../app/ajaxStatusDucks';
import * as mockSignalR from '../../api/mockSignalR';

const prefix = 'checklist-chat/chat/';

// won't be used with SignalR?:
export const LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS = `${prefix}LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS`;

export const DISPLAY_CHAT_MESSAGE = `${prefix}DISPLAY_CHAT_MESSAGE`;
export const DISPLAY_CHAT_USER_ACTION = `${prefix}DISPLAY_CHAT_USER_ACTION`;
export const DISPLAY_PRIORITY_NOTIFICATION = `${prefix}DISPLAY_PRIORITY_NOTIFICATION`;
export const UPDATE_CHAT_PARTICIPANTS = `${prefix}UPDATE_CHAT_PARTICIPANTS`;

// Actions:

export const loadChatMessagesForRoomSuccess =
    (chatMessages) => ({type: LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS, chatMessages});

export const updateChatParticipants =
    (chatParticipants) => ({type: UPDATE_CHAT_PARTICIPANTS, chatParticipants});

const updateMessagesAndParticipants = (chatMessages, dispatch) => {
    dispatch(loadChatMessagesForRoomSuccess(chatMessages));

    const sortedNames = chatMessages.map(m => m.userName).sort();
    const uniqueNames = [...new Set(sortedNames)];

    const chatParticipants = uniqueNames.map(userName => ({
        id: userName,
        name: userName,
        department: 'Department',
        title: 'Title',
        connection: 'Connection'
    }));

    dispatch(updateChatParticipants(chatParticipants));
};

export function loadChatMessagesForRoom(roomId) {
    if (roomId) {
        return dispatch => {
            dispatch(beginAjaxCall());

            return mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch)
            ).catch(error => {
                dispatch(ajaxCallError(error));
                throw (error);
            });
        };
    } else {
        return [];
    }
}

export function saveChatMessage(chatMessage, roomId) {
    return dispatch => {
        dispatch(beginAjaxCall());

        return mockSignalR.chat(roomId, chatMessage).then(() =>
        {
            mockSignalR.getChatMessages(roomId).then(chatMessages =>
                updateMessagesAndParticipants(chatMessages, dispatch));
        }).catch(error => {
            dispatch(ajaxCallError(error));
            throw (error);
        });
    };
}

// Reducers:

export function chatMessagesReducer(chatMessages = initialState.chatMessages, action) {
    const actionType = action.type;

    switch (actionType) {

        case LOAD_CHAT_MESSAGES_FOR_ROOM_SUCCESS:
            return action.chatMessages;

        case DISPLAY_CHAT_MESSAGE:
        case DISPLAY_CHAT_USER_ACTION:
            return [...chatMessages, Object.assign({}, action.chatMessage)];

        default:
            return chatMessages;
    }
}

export function chatParticipantsReducer(chatParticipants = initialState.chatParticipants, action) {
    const actionType = action.type;

    switch (actionType) {

        case UPDATE_CHAT_PARTICIPANTS:
            return action.chatParticipants;

        default:
            return chatParticipants;
    }
}
