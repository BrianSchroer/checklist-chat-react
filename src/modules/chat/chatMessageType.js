export const CHAT = 'Chat';
export const ACTION = 'Action';
