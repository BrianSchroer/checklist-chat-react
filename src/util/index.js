import format from './format';
import uiHelper from './uiHelper';
import validationHelper from './validationHelper';

export {
    format as format,
    uiHelper as uiHelper,
    validationHelper as validationHelper
};
