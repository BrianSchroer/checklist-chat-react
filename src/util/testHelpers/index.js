import {shallow, mount} from 'enzyme';
import enzymeHelper from './enzymeHelper';

export {
    shallow as shallow,
    mount as mount,
    enzymeHelper as enzymeHelper
};
