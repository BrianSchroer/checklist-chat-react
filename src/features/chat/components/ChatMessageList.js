import React, {PropTypes} from 'react';
import ChatMessage from './ChatMessage';
import ChatActionMessage from './ChatActionMessage';
import * as chatMessageType from '../chatMessageType';

const ChatMessageList = ({messages}) => {
    return (
        <ul className="chat-message-list list-group">
            {messages.map(message =>
                (message.chatMessageType === chatMessageType.ACTION)
                    ? <ChatActionMessage key={message.key} message={message}/>
                    : <ChatMessage key={message.key} message={message}/>
            )}
        </ul>
    );
};

ChatMessageList.propTypes = {
    messages: PropTypes.arrayOf(PropTypes.object).isRequired
};

export default ChatMessageList;
