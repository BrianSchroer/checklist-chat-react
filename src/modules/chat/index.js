import * as chatMessageType from './chatMessageType';
import ChatMessageMarkdown from './components/ChatMessageMarkdown';

export {
    chatMessageType as chatMessageType,
    ChatMessageMarkdown as ChatMessageMarkdown
};
