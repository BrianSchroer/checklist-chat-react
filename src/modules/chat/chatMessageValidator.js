import {validationHelper} from '../../util';

export function validate(chatMessage) {
    return validationHelper.validateFields([{
        fieldName: 'text',
        predicate: validationHelper.missingRequiredValue(chatMessage.text),
        errorMessage: 'Message is required.'
    }]);
}
