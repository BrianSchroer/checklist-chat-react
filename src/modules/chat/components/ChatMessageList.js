import React from 'react';
import PropTypes from 'prop-types';
import { uiHelper } from '../../../util';
import ChatMessage from './ChatMessage';

class ChatMessageList extends React.Component {
  componentDidUpdate() {
    if (this.props.shouldScrollToBottom) {
      uiHelper.afterRenderIsComplete(() => {
        uiHelper.scrollToBottom(this.messageList);
      });
    }
  }

  render() {
    const { userId, chatMessages } = this.props;

    return (
      <ul
        className="chat-message-list list-unstyled"
        ref={ul => (this.messageList = ul)} //eslint-disable-line react/jsx-no-bind
      >
        {chatMessages.map(message => (
          <li key={message.id}>
            <ChatMessage userId={userId} chatMessage={message} />
          </li>
        ))}
      </ul>
    );
  }
}

ChatMessageList.propTypes = {
  userId: PropTypes.string.isRequired,
  chatMessages: PropTypes.arrayOf(PropTypes.object).isRequired,
  shouldScrollToBottom: PropTypes.bool
};

ChatMessageList.defaultProps = {
  shouldScrollToBottom: false
};

export default ChatMessageList;
