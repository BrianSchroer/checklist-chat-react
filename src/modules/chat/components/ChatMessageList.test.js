import React from 'react';
import {snapshotHelper} from '../../../util/testHelpers';
import ChatMessageList from './ChatMessageList';
import {chatMessageType} from '../../chat';

describe('ChatMessageList', () => {
    it('should render correctly', () => {
        const props = {
            userId: 'test user id',
            chatMessages: [
                { id: 1, timeStamp: '2016-12-08T14:57:11.222Z', userName: 'user1', text: 'text1', chatMessageType: chatMessageType.CHAT },
                { id: 2, timeStamp: '2016-12-08T14:57:12.222Z', userName: 'user2', text: 'text2', chatMessageType: chatMessageType.ACTION },
                { id: 3, timeStamp: '2016-12-08T14:57:13.222Z', userName: 'user3', text: 'text3', chatMessageType: chatMessageType.CHAT }
            ]
        };

        snapshotHelper.assertMatch(<ChatMessageList {...props} />);
    });
});
