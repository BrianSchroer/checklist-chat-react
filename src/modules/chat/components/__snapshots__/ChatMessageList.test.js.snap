// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageList should render correctly 1`] = `
<ul
  className="chat-message-list list-unstyled"
  id="chatMessageList"
>
  <li>
    <div
      className="chat-message"
    >
      <div
        className="timestamp chat-message-timestamp"
      >
        8:57 am
      </div>
      <strong>
        user1: 
      </strong>
      <div
        className="chat-message-text"
      >
        <p>
          text1
        </p>
      </div>
    </div>
  </li>
  <li>
    <div
      className="chat-action-message text-muted"
    >
      <div
        className="timestamp chat-message-timestamp"
      >
        8:57 am
      </div>
      <span
        className="chat-action-message-text"
      >
        user2 text2
      </span>
    </div>
  </li>
  <li>
    <div
      className="chat-message"
    >
      <div
        className="timestamp chat-message-timestamp"
      >
        8:57 am
      </div>
      <strong>
        user3: 
      </strong>
      <div
        className="chat-message-text"
      >
        <p>
          text3
        </p>
      </div>
    </div>
  </li>
</ul>
`;
